src/neuroglancer/datasource/deepzoom/frontend.ts

```diff
@@ -0,0 +1,310 @@
+import {getObjectId} from 'neuroglancer/util/object_id';
+import {cancellableFetchSpecialOk, parseSpecialUrl, SpecialProtocolCredentials, SpecialProtocolCredentialsProvider} from 'neuroglancer/util/special_protocol_request';
+
+/*export*/ class DeepzoomImageTileSource extends
+(WithParameters(WithCredentialsProvider<SpecialProtocolCredentials>()(VolumeChunkSource), ImageTileSourceParameters)) {}
+
+class ScaleInfo {
+  key: string;
+  encoding: ImageTileEncoding;
+  resolution: Float64Array;
+  voxelOffset: Float32Array;
+  size: Float32Array;
+  chunkSizes: Uint32Array[];
+  constructor(obj: any, numChannels: number) {
+    verifyObject(obj);
+    const rank = (numChannels === 1) ? 3 : 4;
+    const resolution = this.resolution = new Float64Array(rank);
+    const voxelOffset = this.voxelOffset = new Float32Array(rank);
+    const size = this.size = new Float32Array(rank);
+    if (rank === 4) {
+      resolution[3] = 1;
+      size[3] = numChannels;
+    }
+    verifyObjectProperty(
+        obj, 'resolution',
+        x => parseFixedLengthArray(resolution.subarray(0, 3), x, verifyFinitePositiveFloat));
+    verifyOptionalObjectProperty(
+        obj, 'voxel_offset', x => parseFixedLengthArray(voxelOffset.subarray(0, 3), x, verifyInt));
+    verifyObjectProperty(
+        obj, 'size', x => parseFixedLengthArray(size.subarray(0, 3), x, verifyPositiveInt));
+    this.chunkSizes = verifyObjectProperty(
+        obj, 'chunk_sizes', x => parseArray(x, y => {
+                              const chunkSize = new Uint32Array(rank);
+                              if (rank === 4) chunkSize[3] = numChannels;
+                              parseFixedLengthArray(chunkSize.subarray(0, 3), y, verifyPositiveInt);
+                              return chunkSize;
+                            }));
+    if (this.chunkSizes.length === 0) {
+      throw new Error('No chunk sizes specified.');
+    }
+    this.encoding =
+        verifyObjectProperty(obj, 'encoding', x => verifyEnumString(x, ImageTileEncoding));
+    this.key = verifyObjectProperty(obj, 'key', verifyString);
+  }
+}
+
+/*export*/ interface PyramidalImageInfo {
+  dataType: DataType;
+  volumeType: VolumeType;
+  scales: ScaleInfo[];
+  modelSpace: CoordinateSpace;
+  overlap: number;
+  tilesize: number;
+}
+
+/*export*/ function buildPyramidalImageInfo(metadata: DZIMetaData): PyramidalImageInfo {
+  const {width, height, tilesize, overlap, format} = metadata;
+  const dataType = DataType.UINT8;
+  const numChannels = 3;
+  const volumeType = VolumeType.IMAGE;
+  const scaleInfos = new Array<ScaleInfo>();
+  let w = width, h = height;
+  let maxlevel = Math.ceil(Math.log2(Math.max(w, h)));
+  do {
+    const lvl = scaleInfos.length;
+    const res = 1 << lvl;
+    scaleInfos.push(new ScaleInfo(
```

Comment:
Instead of converting to json here and then parsing it in the `ScaleInfo` constructor, just make `ScaleInfo` an interface.

Reply:
Thanks for the suggestion, I will do that. Just haven't got to these parts yet.

Reply:
Actually I burnt myself with this multiple times last week, but now I succeeded with it in much smaller steps.
Done in https://github.com/google/neuroglancer/commit/ceca06e04a75b5eb39eb435dbd3058decefb5d85